pkg/3scale/amp/component/generic_monitoring.go

```diff
@@ -0,0 +1,152 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func KubernetesResourcesByNamespaceGrafanaDashboard(ns string) *grafanav1alpha1.GrafanaDashboard {
+	data := &struct {
+		Namespace string
+	}{
+		ns,
+	}
+	return &grafanav1alpha1.GrafanaDashboard{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "kubernetes-resources-by-namespace",
+			Labels: map[string]string{
+				"monitoring-key": common.MonitoringKey,
+			},
+		},
+		Spec: grafanav1alpha1.GrafanaDashboardSpec{
+			Json: assets.TemplateAsset("monitoring/kubernetes-resources-by-namespace-grafana-dashboard-1.json.tpl", data),
+			Name: fmt.Sprintf("%s/kubernetes-resources-by-namespace-grafana-dashboard-1.json", ns),
+		},
+	}
+}
+
+func KubernetesResourcesByPodGrafanaDashboard(ns string) *grafanav1alpha1.GrafanaDashboard {
+	data := &struct {
+		Namespace string
+	}{
+		ns,
+	}
+	return &grafanav1alpha1.GrafanaDashboard{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "kubernetes-resources-by-pod",
+			Labels: map[string]string{
+				"monitoring-key": common.MonitoringKey,
+			},
+		},
+		Spec: grafanav1alpha1.GrafanaDashboardSpec{
+			Json: assets.TemplateAsset("monitoring/kubernetes-resources-by-pod-grafana-dashboard-1.json.tpl", data),
+			Name: fmt.Sprintf("%s/kubernetes-resources-by-pod-grafana-dashboard-1.json", ns),
+		},
+	}
+}
+
+func KubeStateMetricsPrometheusRules(ns string) *monitoringv1.PrometheusRule {
+	return &monitoringv1.PrometheusRule{
+		ObjectMeta: v12.ObjectMeta{
+			Name: "threescale-kube-state-metrics",
+			Labels: map[string]string{
+				"monitoring-key": common.MonitoringKey,
+				"prometheus":     "application-monitoring",
+				"role":           "alert-rules",
+			},
+		},
+		Spec: monitoringv1.PrometheusRuleSpec{
+			Groups: []monitoringv1.RuleGroup{
+				{
+					Name: fmt.Sprintf("%s/threescale-kube-state-metrics.rules", ns),
+					Rules: []monitoringv1.Rule{
+						{
+							Alert: "ThreescalePodCrashLooping",
+							Annotations: map[string]string{
+								"message": `Pod {{ $labels.namespace }}/{{ $labels.pod }} ({{ $labels.container }}) is restarting {{ printf "%.2f" $value }} times / 5 minutes.`,
+							},
+							Expr: intstr.FromString(fmt.Sprintf(`rate(kube_pod_container_status_restarts_total{namespace="%s",pod=~"(apicast-.*|backend-.*|system-.*|zync-.*)"}[15m]) * 60 * 5 > 0`, ns)),
+							For:  "5m",
+							Labels: map[string]string{
+								"severity": "critical",
+							},
+						},
+						{
+							Alert: "ThreescalePodNotReady",
+							Annotations: map[string]string{
+								"message": `Pod {{ $labels.namespace }}/{{ $labels.pod }} has been in a non-ready state for longer than 5 minutes.`,
+							},
+							Expr: intstr.FromString(fmt.Sprintf(`sum by (namespace, pod) (max by(namespace, pod) (kube_pod_status_phase{namespace="%s",pod=~"(apicast-.*|backend-.*|system-.*|zync-.*)", phase=~"Pending|Unknown"}) * on(namespace, pod) group_left(owner_kind) max by(namespace, pod, owner_kind) (kube_pod_owner{namespace="%s",owner_kind!="Job"})) > 0`, ns, ns)),
```

Comment:
Should we modify existing expr
`sum by (namespace, pod) (max by(namespace, pod) (kube_pod_status_phase{namespace="%s",pod=~"(apicast-.*|backend-.*|system-.*|zync-.*)", phase=~"Pending|Unknown"}) * on(namespace, pod) group_left(owner_kind) max by(namespace, pod, owner_kind) (kube_pod_owner{namespace="%s",owner_kind!="Job"})) > 0`

By 

`sum by (pod) (max by(pod) (kube_pod_status_phase{namespace="%s",pod=~"(apicast-.*|backend-.*|system-.*|zync-.*)", phase=~"Pending|Unknown"}) * on(pod) group_left(owner_kind) max by(pod, owner_kind) (kube_pod_owner{namespace="%s",owner_kind!="Job"})) > 0`

Thus, removing aggregation by namespace??

Reply:
Although the promql query is already filtered by namespace (in order to be able to have a set of alerts with same alert-name, but applying to specific namespaces), the aggregation by namespace is really important, because it adds the label `namespace` to the alert, so you know from which namespace is firing the alert.

Example:
![image](https://user-images.githubusercontent.com/41513123/78661205-a2e90e00-78ce-11ea-8f62-701d5abea44a.png)

Reply:
:+1: